rust-symcrypt/src/ecc/ecdsa.rs

```diff
@@ -0,0 +1,201 @@
+//! EcDsa functions. For further documentation please refer to symcrypt.h
+//! 
+//! # Example
+//! 
+//! ## ECDSA Sign and Verify
+//! 
+//! ```rust
+//! use symcrypt::ecc::{EcKey, CurveType, EcKeyUsage};
+//! use hex::*;
+//! 
+//! // Set up input hash value
+//! let hash_value = hex::decode("4d55c99ef6bd54621662c3d110c3cb627c03d6311393b264ab97b90a4b15214a5593ba2510a53d63fb34be251facb697c973e11b665cb7920f1684b0031b4dd370cb927ca7168b0bf8ad285e05e9e31e34bc24024739fdc10b78586f29eff94412034e3b606ed850ec2c1900e8e68151fc4aee5adebb066eb6da4eaa5681378e").unwrap();
+//! 
+//! // Generate a new ECDSA key pair
+//! let key = EcKey::generate_key_pair(CurveType::NistP256, EcKeyUsage::EcDsa).unwrap();
+//! 
+//! // Sign the hash value
+//! let signature = key.ecdsa_sign(&hash_value).unwrap();
+//! 
+//! // Verify the signature
+//! let result = key.ecdsa_verify(&signature, &hash_value);
+//! 
+//! // Assert the signature is valid
+//! assert!(result.is_ok());
+//! 
+//! ```
+//! 
+use crate::ecc::{EcKey, curve_to_num_format, EcKeyUsage};
+use crate::errors::SymCryptError;
+use std::vec;
+use symcrypt_sys;
+
+impl EcKey {
+    /// `ecdsa_sign()` returns a signature as a `Vec<u8>`, or a [`SymCryptError`] if the operation fails.
+    ///
+    /// `hash_value` is a `&[u8]` that represents the hash value to sign.
+    /// 
+    /// If the key usage is not [`EcKeyUsage::EcDsa`], or [`EcKeyUsage::EcDhAndEcDsa`] the function will
+    /// fail with a [`SymCryptError`] with the value [`SymCryptError::InvalidArgument`].
+    pub fn ecdsa_sign(
+        &self,
+        hash_value: &[u8],
+    ) -> Result<Vec<u8>, SymCryptError> {
+        // SymCrypt code AV's with SymCrypt Fatal Error.
+        // Panic'ing is not normal here in Rust so we are handling the error instead and returning SymCryptError::InvalidArgument
+        if self.get_ec_curve_usage() == EcKeyUsage::EcDh {
+            return Err(SymCryptError::InvalidArgument);
+        } 
+
+        // Per SymCrypt docs, the size of the signature will be 2 x the size of the private key.
+        let signature_size = self.get_size_of_private_key() * 2; 
+        let mut signature = vec![0u8; signature_size as usize]; // must be 2x size of private key SymCryptEckeySizeofPrivateKey
+        unsafe {
+            // SAFETY: FFI calls
+            match symcrypt_sys::SymCryptEcDsaSign(

```

Comment:
>SymCryptEcDsaSign [](http://example.com/codeflow?start=32&length=17)

What does SymCrypt do for an ECDSA key that only has the public key set?

Reply:
This is the test for it, trying to sign without a public key set will result in an Invalid Arg error. 
```
fn test_ecdsa_sign_without_private_key() {
        let key = EcKey::generate_key_pair(CurveType::NistP256, EcKeyUsage::EcDsa).unwrap();
        let hash_value = hex::decode("4d55c99ef6bd54621662c3d110c3cb627c03d6311393b264ab97b90a4b15214a5593ba2510a53d63fb34be251facb697c973e11b665cb7920f1684b0031b4dd370cb927ca7168b0bf8ad285e05e9e31e34bc24024739fdc10b78586f29eff94412034e3b606ed850ec2c1900e8e68151fc4aee5adebb066eb6da4eaa5681378e").unwrap();
        let public_key_bytes = key.export_public_key().unwrap();
        let key = EcKey::set_public_key(CurveType::NistP256, &public_key_bytes, EcKeyUsage::EcDsa).unwrap();
        let result = key.ecdsa_sign(&hash_value);
        assert_eq!(result, Err(SymCryptError::InvalidArgument));
    }
```